src/checksum/software/checksum.c

```diff
@@ -1128,41 +1124,85 @@ __wt_checksum_sw(const void *chunk, size_t len)
-			g_crc_slicing[4][(crc >> 24)] ^
-			g_crc_slicing[3][(next      ) & 0xFF] ^
-			g_crc_slicing[2][(next >>  8) & 0xFF] ^
-			g_crc_slicing[1][(next >> 16) & 0xFF] ^
-			g_crc_slicing[0][(next >> 24)];
-#endif
+		    g_crc_slicing_be[4][(crc      ) & 0xFF] ^
+		    g_crc_slicing_be[5][(crc >>  8) & 0xFF] ^
+		    g_crc_slicing_be[6][(crc >> 16) & 0xFF] ^
+		    g_crc_slicing_be[7][(crc >> 24)] ^
+		    g_crc_slicing_be[0][(next      ) & 0xFF] ^
+		    g_crc_slicing_be[1][(next >>  8) & 0xFF] ^
+		    g_crc_slicing_be[2][(next >> 16) & 0xFF] ^
+		    g_crc_slicing_be[3][(next >> 24)];
 	}
 
 	/* Checksum trailing bytes one byte at a time. */
-#ifdef WORDS_BIGENDIAN
 	for (len &= 0x7; len > 0; ++p, len--)
-		crc = g_crc_slicing[0][((crc >> 24) ^ *p) & 0xFF] ^ (crc << 8);
+		crc =
+		    g_crc_slicing_be[0][((crc >> 24) ^ *p) & 0xFF] ^ (crc << 8);
 
 	/* Do final byte swap to produce a result identical to little endian */
 	crc =
-		((crc << 24) & 0xFF000000) |
-		((crc <<  8) & 0x00FF0000) |
-		((crc >>  8) & 0x0000FF00) |
-		((crc >> 24) & 0x000000FF);
-#else
+	    ((crc << 24) & 0xFF000000) |
+	    ((crc <<  8) & 0x00FF0000) |
+	    ((crc >>  8) & 0x0000FF00) |
+	    ((crc >> 24) & 0x000000FF);
+	return (~crc);
+}
+
+/*
+ * __checksum_sw_le --
+ *	Return a checksum for a chunk of memory, computed in software,
+ * little-endian version.
+ */
+static uint32_t
+__checksum_sw_le(const void *chunk, size_t len)
+{
+	uint32_t crc, next;
+	size_t nqwords;
+	const uint8_t *p;
+
+	crc = 0xffffffff;
+
+	/* Checksum one byte at a time to the first 4B boundary. */
+	for (p = chunk;
+	    ((uintptr_t)p & (sizeof(uint32_t) - 1)) != 0 &&
+	    len > 0; ++p, --len)
+		crc = g_crc_slicing_le[0][(crc ^ *p) & 0xFF] ^ (crc >> 8);
+
+	/* Checksum in 8B chunks. */
+	for (nqwords = len / sizeof(uint64_t); nqwords; nqwords--) {
+		crc ^= *(uint32_t *)p;
+		p += sizeof(uint32_t);
+		next = *(uint32_t *)p;
+		p += sizeof(uint32_t);
+		crc =
+		    g_crc_slicing_le[7][(crc      ) & 0xFF] ^
+		    g_crc_slicing_le[6][(crc >>  8) & 0xFF] ^
+		    g_crc_slicing_le[5][(crc >> 16) & 0xFF] ^
+		    g_crc_slicing_le[4][(crc >> 24)] ^
+		    g_crc_slicing_le[3][(next      ) & 0xFF] ^
+		    g_crc_slicing_le[2][(next >>  8) & 0xFF] ^
+		    g_crc_slicing_le[1][(next >> 16) & 0xFF] ^
+		    g_crc_slicing_le[0][(next >> 24)];
+	}
+
+	/* Checksum trailing bytes one byte at a time. */
 	for (len &= 0x7; len > 0; ++p, len--)
-		crc = g_crc_slicing[0][(crc ^ *p) & 0xFF] ^ (crc >> 8);
-#endif
+		crc = g_crc_slicing_le[0][(crc ^ *p) & 0xFF] ^ (crc >> 8);
+
 	return (~crc);
 }
+
+extern uint32_t (*__wt_checksum_sw(void))(const void *, size_t);
+
+/*
+ * __wt_checksum_sw --
+ *	Determine if on big- or little-endian system and return a checksum for
+ * a chunk of memory, computed in software.
+ */
+uint32_t (*__wt_checksum_sw(void))(const void *, size_t)
+{
+	uint64_t v;
+
+	v = 1;
+	return (*((uint8_t *)&v) == 0 ? __checksum_sw_be : __checksum_sw_le);
```

Comment:
Does the fact that we're checking at runtime mean we can't rely on `WORDS_BIGENDIAN`?  Can we not figure this out during configure because of cross-compilation?

Reply:
I had the same thought, Keith answered in his commit message:

```
MongoDB had a bug, they depended on the autoconfig #define WORDS_BIGENDIAN,
which is no longer included by the crc32c code because MongoDB doesn't
necessarily have them available.
```

Reply:
On the other hand - I am a bit nervous about this check - it feels like something the compiler could break. Is it a known idiom?

Reply:
> Does the fact that we're checking at runtime mean we can't rely on `WORDS_BIGENDIAN`?  Can we not figure this out during configure because of cross-compilation?

This code can't depend on any of the normal autoconfigure work because it's separately compiled for MongoDB services when we may not be building a WiredTiger library.

Reply:
> On the other hand - I am a bit nervous about this check - it feels like something the compiler could break. Is it a known idiom?

I stared at the standard and searched around some, and this seems as safe/portable as anything I could find. It seems to me that as long as a `uint64_t` is larger than a `uint8_t`, it has to work (famous last words).

We use that same check in `support/global.c:__wt_endian_check()` to detect if the build being run matches the machine on which we're running, and that at least tells us that every compiler we've seen so far has no complaints.

Reply:
And then, there's this: [Finally, one-line endianness detection in the C preprocessor](http://esr.ibiblio.org/?p=5095)